curp/src/server/curp_node.rs

```diff
@@ -702,28 +714,43 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
                 }
             }
 
-            let Some(sync_action) = curp.sync(id) else {
+            let Some(sync_action) = curp.sync(connect_id) else {
                 break true;
             };
 
             match sync_action {
                 SyncAction::AppendEntries(ae) => {
                     let is_empty = ae.entries.is_empty();
+                    let is_commit_shutdown = ae.entries.last().is_some_and(|e| {
+                        matches!(e.entry_data, EntryData::Shutdown(_))
+                            && e.index == ae.leader_commit
+                    });
                     // (hb_opt, entries) status combination
                     // (false, empty) => send heartbeat to followers
                     // (true, empty) => indicates that `batch_timeout` expired, and during this period there is not any log generated. Do nothing
                     // (true | false, not empty) => send append entries
                     if !hb_opt || !is_empty {
                         let result = Self::send_ae(connect.as_ref(), curp.as_ref(), ae).await;
                         if let Err(err) = result {
-                            warn!("ae to {} failed, {err}", connect.id());
+                            warn!("ae to {} failed, {err}", connect_id);
                             if matches!(err, SendAEError::NotLeader) {
                                 break true;
                             }
                         } else {
                             hb_opt = true;
                         }
-                        if is_shutdown_state && is_empty && curp.is_synced() {
+                        // Check Follower shutdown
+                        if is_shutdown_state
+                            && ((curp.is_synced(connect_id) && is_empty)
+                                || (!curp.is_synced(connect_id) && is_commit_shutdown))
+                        {
+                            let _ig = tokio::spawn(async move {
```

Comment:
This is a async task, could the leader has already shutdowned before these task still not get started?